File: trunk/ninjac/loops.cpp
/*
 *      NINJAC - an interative, programmable calculator
 *
 *      semestral project for C/C++ programming course
 *      (Y36PJC) at the FEE CTU Prague
 *
 *      Created by Pavel Sramek (sramepa1@fel.cvut.cz)
 *      December 2009
 *
 *      This is free software, licensed under GNU LGPL
 *      (GNU Lesser General Public License, version 3)
 *      http://www.gnu.org/licenses/lgpl.html
 *
 *      Project homepage:
 *      http://code.google.com/p/ninjac/
 *
 *      Version 1.0
 *
 */
#ifdef DEBUG    
    #include <assert.h>    
#endif

#include <iostream>
#include <cmath>

#include "loops.h"
#include "globals.h"

using namespace std;

ForLoop::~ForLoop() {
    if(fromExpr != NULL) delete fromExpr;
    if(toExpr != NULL) delete toExpr;
    if(stepExpr != NULL) delete stepExpr;
}

/*
 * executes a for-loop. Algorithm is:
 * - evaluate bounds and step
 * - determine direction
 * - repeat execution and control var assignment until second bound is stepped over
 */
void ForLoop::execute() {
    #ifdef DEBUG
        cout << "### entering for loop, evaluating expressions" << endl;
        assert(fromExpr);
        assert(toExpr);
        assert(stepExpr);
    #endif

    double from = fromExpr->evaluate();
    double to = toExpr->evaluate();
    double step = fabs(stepExpr->evaluate()); // strips sign
    bool up = to >= from;

    #ifdef DEBUG
        cout << "### for loop control var name: " << vName << endl
             << "###   from=" << from << " to=" << to << " step=" << step << endl;
    #endif

    double varValue = from;
    while(  (up && (varValue <= to + Globals::inst->delta))
          || (!up && (varValue >= to - Globals::inst->delta))) {

        #ifdef DEBUG
            cout << "### executing an iteration of for loop, $" << vName << " is " << varValue << ',' << endl
                 << "###   iteration is from " << from << " to " << to << " step " << step << endl;
        #endif

        Globals::inst->assignVar(vName,varValue);
        stmt->execute();
        varValue = up ? varValue + step : varValue - step;
    }

    if((varValue == from)) { // unchanged
        Globals::inst->assignVar(vName,from);
    }

    if(Globals::inst->ia() && topLevel) {
        cout << "#> OK" << endl;
    }

    #ifdef DEBUG
        cout << "### leaving for loop, $" << vName << '=' << Variable(vName).evaluate() << " ( " << (up?">= ":"<= ") << to << " )" << endl;
    #endif
}


/*
 * repeats body execution until condition is true
 */
void RepeatLoop::execute() {
    #ifdef DEBUG
        cout << "### entering repeat loop" << endl;
    #endif

    do {
        #ifdef DEBUG
            cout << "### executing an iteration of repeat loop" << endl;
        #endif

        stmt->execute();

    }while (fabs(cond->evaluate()) <= Globals::inst->delta);

    if(Globals::inst->ia() && topLevel) {
        cout << "#> OK" << endl;
    }

    #ifdef DEBUG
        cout << "### leaving repeat loop" << endl;
    #endif
}


/*
 * while condition is true, repeats body execution
 */
void WhileLoop::execute() {
    #ifdef DEBUG
        cout << "### entering while loop" << endl;
    #endif

    while (fabs(cond->evaluate()) > Globals::inst->delta) {

        #ifdef DEBUG
            cout << "### executing an iteration of while loop" << endl;
        #endif

        stmt->execute();
    }

    if(Globals::inst->ia() && topLevel) {
        cout << "#> OK" << endl;
    }

    #ifdef DEBUG
        cout << "### leaving while loop" << endl;
    #endif
}

File: web/App_Code/Loops.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

public abstract class Loop : Statement
{
    protected Ninjac n;
    protected bool topLevel;

    public Statement stmt { get; set; }

    public Loop(bool top, Ninjac ninjac)
    {
        topLevel = top;
        n = ninjac;
        stmt = null;
    }

    public abstract void execute();
}


public class ForLoop : Loop
{
    protected string vName;

    public Expression fromExpr { get; set; }
    public Expression toExpr { get; set; }
    public Expression stepExpr { get; set; }

    public ForLoop(string varName, bool top, Ninjac n)
        : base(top, n)
    {
        vName = varName;
    }

    public override void execute()
    {
        double from = fromExpr.evaluate();
        double to = toExpr.evaluate();
        double step = Math.Abs(stepExpr.evaluate());
        bool up = to >= from;

        double varValue = from;
        while ((up && (varValue <= to + Ninjac.DELTA))
              || (!up && (varValue >= to - Ninjac.DELTA)))
        {

            n.assignVar(vName, varValue);
            stmt.execute();
            varValue = up ? varValue + step : varValue - step;
        }

        if ((varValue == from))
        {
            n.assignVar(vName, from);
        }

        if (n.ia() && topLevel)
        {
            n.outputLine("OK");
        }
    }
}


public abstract class CondLoop : Loop
{
    public Expression cond { get; set; }

    public CondLoop(bool top, Ninjac n)
        : base(top, n)
    {
        cond = null;
    }

    //public abstract void execute();
}


public class RepeatLoop : CondLoop
{
    public RepeatLoop(bool top, Ninjac n)
        : base(top, n)
    { }

    public override void execute()
    {
        do
        {
            stmt.execute();
        } while (Math.Abs(cond.evaluate()) <= Ninjac.DELTA);

        if (n.ia() && topLevel)
        {
            n.outputLine("OK");
        }
    }
}


public class WhileLoop : CondLoop
{
    public WhileLoop(bool top, Ninjac n)
        : base(top, n)
    { }

    public override void execute()
    {
        while (Math.Abs(cond.evaluate()) > Ninjac.DELTA)
        {
            stmt.execute();
        }

        if (n.ia() && topLevel)
        {
            n.outputLine("OK");
        }
    }
}


